balancer/rls/balancer_test.go

```diff
@@ -650,6 +650,181 @@ func (s) TestConfigUpdate_DataCacheSizeDecrease(t *testing.T) {
+	rlsServer.SetResponseCallback(func(ctx context.Context, req *rlspb.RouteLookupRequest) *rlstest.RouteLookupResponse {
+		if req.KeyMap["k1"] == "v1" {
+			return &rlstest.RouteLookupResponse{Resp: &rlspb.RouteLookupResponse{Targets: []string{backendAddress1}}}
+		}
+		if req.KeyMap["k2"] == "v2" {
+			return &rlstest.RouteLookupResponse{Resp: &rlspb.RouteLookupResponse{Targets: []string{backendAddress2}}}
+		}
+		return &rlstest.RouteLookupResponse{Err: errors.New("no keys in request metadata")}
+	})
+
+	// Register a manual resolver and push the RLS service config through it.
+	r := manual.NewBuilderWithScheme("rls-e2e")
+	headers := `
+    [
+        {
+            "key": "k1",
+            "names": [
+                "n1"
+            ]
+        },
+        {
+            "key": "k2",
+            "names": [
+                "n2"
+            ]
+        }
+    ]
+    `
+
+	configJSON := `
+	{
+	  "loadBalancingConfig": [
+		{
+		  "%s": {
+			"routeLookupConfig": {
+				"grpcKeybuilders": [{
+					"names": [{"service": "grpc.testing.TestService"}],
+					"headers": %s
+				}],
+				"lookupService": "%s",
+				"cacheSizeBytes": %d
+			},
+			"childPolicy": [{"%s": {}}],
+			"childPolicyConfigTargetFieldName": "Backend"
+		  }
+		}
+	  ]
+	}`
+	scJSON := fmt.Sprintf(configJSON, topLevelBalancerName, headers, rlsServer.Address, 1000, childPolicyName)
+	sc := internal.ParseServiceConfig.(func(string) *serviceconfig.ParseResult)(scJSON)
+	r.InitialState(resolver.State{ServiceConfig: sc})
+
+	ccCh := make(chan *grpc.ClientConn, 1)
+	go func() {
+		cc, err := grpc.Dial(r.Scheme()+":///", grpc.WithResolvers(r), grpc.WithTransportCredentials(insecure.NewCredentials()))
+		if err != nil {
+			return
+		}
+		ccCh <- cc
+	}()
+	<-clientConnUpdateDone
+	cc := <-ccCh
+	defer cc.Close()
+
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	makeTestRPCAndVerifyError(ctx, t, cc, codes.Unavailable, nil)
+	t.Logf("Verifying if RPC failed when listener is stopped.")
+
+	ctxOutgoing := metadata.AppendToOutgoingContext(ctx, "n1", "v1")
+	makeTestRPCAndExpectItToReachBackend(ctxOutgoing, t, cc, backendCh1)
+	verifyRLSRequest(t, rlsReqCh, true)
+
+	ctxOutgoing = metadata.AppendToOutgoingContext(ctx, "n2", "v2")
+	makeTestRPCAndExpectItToReachBackend(ctxOutgoing, t, cc, backendCh2)
+	verifyRLSRequest(t, rlsReqCh, true)
+
+	// Override the newPickerGenerated to measure the number of times
+	// the picker is generated because state updates can be inhibited.
+	//
+	// Note: This needs to be initialised here and not in the beginning
+	// of the test, as otherwise for every RPC call we make, it'd have
+	// affected total picker count.
+	pckrSentBeforeClientConnUpdate := make(chan struct{}, 1)
+	totalPickerGenerated := 0
+	origNewPickerGenerated := newPickerGenerated
+	newPickerGenerated = func() {
+		totalPickerGenerated++
+		if totalPickerGenerated == 2 {
+			pckrSentBeforeClientConnUpdate <- struct{}{}
+		}
+	}
```

Comment:
Instead of maintaining a counter and and a channel, this can be done using a blocking channel.
```go
pickerUpdateChan := make(chan struct{})
newPickerGenerated = func() {
	newPickerGenerated<-struct{}{}
}

// snipped
for i := 0; i < 2; i++{
	select {
	case <-pckrSentBeforeClientConnUpdate:
	case <-clientConnUpdateDone:
		t.Fatalf("Client conn update was completed before picker update.")
	case <-ctx.Done():
		t.Fatalf("client conn update could not complete: %v", ctx.Err().Error())
	}
}
```

Reply:
Actually that might not be correct as we are making the RPC calls in between, so that could also make the picker updated. For ease of the test I think the current state of PR makes more sense to me, as it specifically verifies for status updates that there should be only 2 picker updates happening on cache resizing. WDYT?

Reply:
Usually test setup happens before the test is run and results are verified. 
If you want to prevent the blocking till the last update, you can use an atomic boolean like `blockUpdates := atomic.Bool{}`.

In the picker update hook, you can check the bool and let the request through. Just before calling `UpdateState`, you can set the boolean to true.

Reply:
Also reading from the blocking channel twice is a cleaner way than having the hook send one message on the channel after 2 calls.